packages/fast-components-react-msft/src/text-action/text-action.tsx

```diff
@@ -0,0 +1,167 @@
+
+        this.state = {
+            focused: false,
+        };
+    }
+
+    /**
+     * Renders the component
+     */
+    public render(): JSX.Element {
+        return (
+            <div className={this.generateClassNames()}>
+                {this.buttonExists() &&
+                this.props.buttonPosition === TextActionButtonPosition.before
+                    ? this.generateButton()
+                    : null}
+                {this.generateBeforeGlyph()}
+                <TextField
+                    {...this.unhandledProps()}
+                    disabled={this.props.disabled}
+                    placeholder={this.props.placeholder}
+                    jssStyleSheet={textFieldOverrides}
+                    onBlur={this.handleOnBlur}
+                    onFocus={this.handleOnFocus}
+                />
+                {this.generateAfterGlyph()}
+                {this.buttonExists() &&
+                this.props.buttonPosition === TextActionButtonPosition.after
+                    ? this.generateButton()
+                    : null}
+            </div>
+        );
+    }
+
+    /**
+     * Generates class names
+     */
+    protected generateClassNames(): string {
+        let classNames: string = get(this.props, "managedClasses.textAction");
+
+        if (this.props.disabled) {
+            classNames = `${classNames} ${get(
+                this.props,
+                "managedClasses.textAction__disabled"
+            )}`;
+        }
+
+        if (this.state.focused) {
+            classNames = `${classNames} ${get(
+                this.props,
+                "managedClasses.textAction__focus"
+            )}`;
+        }
+
+        return super.generateClassNames(classNames);
+    }
+
+    /**
+     * Adds focus state to outer wrapper
+     * In order to correctly focus the input and then the
+     * possible button, a class must be added instead of using
+     * focus-within via style
+     */
+    private handleOnFocus = (): void => {
+        this.setState({ focused: true });
+    };
+
+    /**
+     * Removes focus state
+     */
+    private handleOnBlur = (): void => {
+        this.setState({ focused: false });
+    };
+
+    /**
+     * Returns truthy if button exist
+     */
+    private buttonExists(): boolean {
+        return typeof this.props.button === "function";
+    }
+
+    /**
+     * Generate button
+     */
+    private generateButton(): React.ReactNode {
+        return this.props.button(
+            get(this.props, "managedClasses.textAction_button"),
+            this.props.disabled
+        );
+    }
+
+    /**
+     * Generates after glyph based on props
+     */
+    private generateAfterGlyph(): React.ReactNode {
+        if (typeof this.props.afterGlyph === "function") {
+            if (
+                (this.buttonExists() &&
+                    this.props.buttonPosition !== TextActionButtonPosition.after) ||
+                !this.buttonExists()
+            ) {
```

Comment:
I think we can lose some of the verbosity here if we check if there is no button first, right? If that check is true, we can perform the next check under that assumption.

```
if (!this.buttonExists() || this.props.buttonPosition !== TextActionButtonPosition.before) {}
```